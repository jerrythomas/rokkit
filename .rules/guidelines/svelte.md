# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
