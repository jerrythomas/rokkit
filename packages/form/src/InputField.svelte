<script>
	import { pick, omit } from 'ramda'
	import { Input } from '@svelte-spice/input'

	let className = ''
	export { className as class }
	export let name
	export let label = null
	export let icon = null
	export let value
	export let type = 'text'
	export let required = false
	export let status = 'default'
	export let disabled = false
	export let message = null

	$: pass = status === 'pass'
	$: fail = status === 'fail'
	$: warn = status === 'warn'
	$: rootProps = pick(['id'], $$restProps)
	$: props = {
		required,
		readOnly: disabled,
		...omit(['id'], $$restProps)
	}
</script>

<input-field
	{...rootProps}
	class="flex flex-col {className}"
	class:disabled
	class:pass
	class:fail
	class:warn
>
	<!-- svelte-ignore a11y-label-has-associated-control -->
	<label class="flex flex-col">
		{#if label}
			<p>{label}</p>
		{/if}
		{#if icon}
			<field class="flex flex-row w-full">
				<span class="flex aspect-square items-center justify-center h-full">
					<icon class={icon} />
				</span>
				<Input bind:value {name} {type} {...props} />
			</field>
		{:else}
			<Input bind:value {name} {type} {...props} />
		{/if}
	</label>
	{#if message}
		<message class={status}>{message}</message>
	{/if}
</input-field>
