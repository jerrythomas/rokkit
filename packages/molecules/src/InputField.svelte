<script>
	import { pick, omit } from 'ramda'
	import { Input, Icon } from '@rokkit/atoms'

	let className = ''
	export { className as class }
	export let name
	export let label = null
	export let description = null
	export let icon = null
	export let value
	export let type = 'text'
	export let required = false
	export let status = 'default'
	export let disabled = false
	export let message = null

	$: pass = status === 'pass'
	$: fail = status === 'fail'
	$: warn = status === 'warn'
	$: rootProps = pick(['id'], $$restProps)
	$: props = {
		required,
		readOnly: disabled,
		...omit(['id'], $$restProps)
	}
</script>

<input-field
	{...rootProps}
	class="flex flex-col input-{type} {className} "
	class:disabled
	class:pass
	class:fail
	class:warn
	class:empty={!value}
>
  {#if label}
	<label for={name} class:required>
		{label}
	</label>
	{/if}
	<field class="w-full flex flex-row items-center">
		{#if icon}
			<Icon name={icon} />
		{/if}
		<Input id={name} bind:value {name} {type} {...props} aria-label={description ?? label ?? name}/>
	</field>
	{#if message}
		<message class={status}>{message}</message>
	{/if}
</input-field>
