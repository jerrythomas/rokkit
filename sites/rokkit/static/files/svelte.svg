<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M18.2301 11.1217C21.1087 6.47773 15.3383 1.6935 11.2372 4.37297L7.337 6.85957C5.44834 8.08961 4.62832 10.453 5.78004 12.8817C2.90038 17.2501 8.50365 22.3655 12.7729 19.6332L16.6732 17.1466C18.7952 15.7494 19.2821 13.3398 18.2301 11.1217V11.1217ZM8.46536 8.46761L12.3599 5.98242C14.9184 4.45077 17.8644 7.12896 16.8023 9.62435C16.1715 9.12318 15.4685 8.81503 14.9345 8.67627C15.1375 7.87006 14.2016 7.27202 13.5507 7.68841L9.64763 10.1651C8.68258 10.8321 9.71148 12.204 10.6653 11.6284L12.1527 10.6803C15.4905 8.79676 18.6495 13.3192 15.5448 15.5357L11.6502 18.0209C9.30336 19.5215 6.10922 17.1175 7.20785 14.379C7.79145 14.8207 8.41131 15.1456 9.07564 15.3271C8.92896 16.2531 9.86333 16.7033 10.4694 16.3234L14.3626 13.8382C15.3233 13.1178 14.292 11.814 13.3464 12.375L11.8533 13.3231C8.55362 15.2514 5.32106 10.6843 8.46536 8.46761Z" fill="black"/>
</svg>
