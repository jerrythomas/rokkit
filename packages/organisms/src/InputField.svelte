<script>
	import { getContext } from 'svelte'
	import { pick, omit } from 'ramda'
	import { Icon } from '@rokkit/atoms'
	import Input from './input/Input.svelte'
	import { componentTypes } from './types'

	const registry = getContext('registry')

	let className = ''
	export { className as class }
	export let name
	export let value = null
	export let label = null
	export let description = null
	export let icon = null
	export let type = 'text'
	export let required = false
	export let status = 'default'
	export let disabled = false
	export let message = null
	export let using = {}
	export let nolabel = false

	$: using = { ...componentTypes, ...$registry, ...using }
	$: pass = status === 'pass'
	$: fail = status === 'fail'
	$: warn = status === 'warn'
	$: rootProps = pick(['id'], $$restProps)
	$: props = {
		required,
		readOnly: disabled,
		...omit(['id'], $$restProps),
		name
	}
</script>

<input-field
	{...rootProps}
	class="flex flex-col input-{type} {className} "
	class:disabled
	class:pass
	class:fail
	class:warn
	class:empty={!value}
>
	{#if label && !nolabel}
		<label for={name} class:required>
			{label}
		</label>
	{/if}
	<field class="w-full flex flex-row items-center" aria-label={description ?? label ?? name}>
		{#if icon}
			<Icon name={icon} />
		{/if}
		<Input id={name} bind:value {type} {...props} {using} />
	</field>
	{#if message}
		<message class={status}>{message}</message>
	{/if}
</input-field>
