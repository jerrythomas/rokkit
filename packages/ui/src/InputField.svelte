<script>
	import { getContext } from 'svelte'
	import { pick, omit } from 'ramda'
	import Icon from './Icon.svelte'
	import Input from './input/Input.svelte'
	import { componentTypes } from './types'

	const registry = getContext('registry')

	let {
		class: className,
		name,
		value,
		type,
		required,
		status,
		disabled,
		message,
		using,
		nolabel,
		icon,
		label,
		description,
		...restProps
	} = $props()

	using = { ...componentTypes, ...registry, ...using }
	let pass = status === 'pass'
	let fail = status === 'fail'
	let warn = status === 'warn'
	let rootProps = pick(['id'], restProps)
	let properties = {
		required,
		readOnly: disabled,
		...omit(['id'], restProps),
		name
	}
</script>

<input-field
	{...rootProps}
	class="flex flex-col input-{type} {className} "
	class:disabled
	class:pass
	class:fail
	class:warn
	class:empty={!value}
>
	{#if label && !nolabel && !['switch', 'checkbox'].includes(type)}
		<label for={name} class:required>
			{label}
		</label>
	{/if}
	<field class="flex w-full flex-row items-center" aria-label={description ?? label ?? name}>
		{#if icon}
			<Icon name={icon} />
		{/if}
		{#if type === 'switch'}
			<label for={name} class:required>{label}</label>
		{/if}
		<Input id={name} bind:value {type} {...properties} {using} on:change />
		{#if type === 'checkbox'}
			<label for={name} class:required>{label}</label>
		{/if}
	</field>
	{#if message}
		<message class={status}>{message}</message>
	{/if}
</input-field>
