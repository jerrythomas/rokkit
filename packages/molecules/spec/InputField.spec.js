import { describe, expect, it, beforeEach } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import { tick } from 'svelte'
import InputField from '../src/InputField.svelte'

describe('InputField.svelte', () => {
	beforeEach(() => cleanup())

	it('should render default', async () => {
		const { container, component } = render(InputField, { name: 'name' })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		let input = container.querySelector('input')
		expect(input).toBeTruthy()
		expect(input.value).toBe('')

		// handle value change
		component.$set({ value: 'hello' })
		await tick()
		input = container.querySelector('input')
		expect(input).toBeTruthy()
		expect(input.value).toBe('hello')
	})

	it('should render null', () => {
		const { container } = render(InputField, { value: null })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render with value', () => {
		const { container } = render(InputField, { value: 'text' })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render with label and value', async () => {
		const { container, component } = render(InputField, {
			label: 'text',
			value: 'text'
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
		// handle label change
		component.$set({ label: 'other' })
		await tick()
		expect(container).toMatchSnapshot()
	})

	it.each(['pass', 'fail', 'warn'])(
		'should render with label, value and status=%s',
		(status) => {
			const { container } = render(InputField, {
				label: 'text',
				value: 'text',
				status,
				message: 'error'
			})
			expect(container).toBeTruthy()
			expect(container).toMatchSnapshot()
		}
	)

	it('should render with description', () => {
		const { container } = render(InputField, {
			label: 'text',
			value: 'text',
			description: 'description'
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render when disabled', async () => {
		const { container, component } = render(InputField, {
			label: 'text',
			value: 'text',
			disabled: true
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
		// handle disabled state change
		component.$set({ disabled: false })
		await tick()
		expect(container).toMatchSnapshot()
	})

	it('should render with custom class an type', async () => {
		const { container, component } = render(InputField, {
			type: 'number',
			class: 'custom'
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		// handle class change
		component.$set({ class: 'other' })
		await tick()
		let wrapper = container.querySelector('input-field')
		expect(wrapper).toBeTruthy()
		expect(Array.from(wrapper.classList)).toContain('other')
		expect(Array.from(wrapper.classList)).toContain('input-number')
	})

	it('should render with icon', async () => {
		const { container, component } = render(InputField, {
			icon: 'icon'
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		// handle icon change
		component.$set({ icon: 'other' })
		await tick()
		let icon = container.querySelector('i')
		expect(icon).toBeTruthy()
		expect(Array.from(icon.classList)).toContain('other')
	})
})
