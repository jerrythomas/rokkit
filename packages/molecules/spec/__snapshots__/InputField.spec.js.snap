// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputField.svelte > should render default 1`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text empty"
    >
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          aria-label="name"
          id="name"
          name="name"
          type="text"
        />
        
        
      </field>
       
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render null 1`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text empty"
    >
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          type="text"
        />
        
        
      </field>
       
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render when disabled 1`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text disabled"
    >
      <label>
        text
      </label>
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          aria-label="text"
          readonly=""
          type="text"
        />
        
        
      </field>
       
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render when disabled 2`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text"
    >
      <label>
        text
      </label>
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          aria-label="text"
          type="text"
        />
        
        
      </field>
       
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render with custom class an type 1`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-number custom empty"
    >
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          type="number"
        />
        
        
      </field>
       
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render with description 1`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text "
    >
      <label>
        text
      </label>
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          aria-label="description"
          type="text"
        />
        
        
      </field>
       
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render with icon 1`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text empty"
    >
       
      <field
        class="w-full flex flex-row items-center"
      >
        <icon
          aria-label="icon"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="icon"
          />
        </icon>
         
        <input
          type="text"
        />
        
        
      </field>
       
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render with label and value 1`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text "
    >
      <label>
        text
      </label>
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          aria-label="text"
          type="text"
        />
        
        
      </field>
       
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render with label and value 2`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text "
    >
      <label>
        other
      </label>
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          aria-label="other"
          type="text"
        />
        
        
      </field>
       
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render with label, value and status=fail 1`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text fail"
    >
      <label>
        text
      </label>
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          aria-label="text"
          type="text"
        />
        
        
      </field>
       
      <message
        class="fail"
      >
        error
      </message>
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render with label, value and status=pass 1`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text pass"
    >
      <label>
        text
      </label>
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          aria-label="text"
          type="text"
        />
        
        
      </field>
       
      <message
        class="pass"
      >
        error
      </message>
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render with label, value and status=warn 1`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text warn"
    >
      <label>
        text
      </label>
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          aria-label="text"
          type="text"
        />
        
        
      </field>
       
      <message
        class="warn"
      >
        error
      </message>
    </input-field>
  </div>
</body>
`;

exports[`InputField.svelte > should render with value 1`] = `
<body>
  <div>
    <input-field
      class="flex flex-col input-text "
    >
       
      <field
        class="w-full flex flex-row items-center"
      >
         
        <input
          type="text"
        />
        
        
      </field>
       
    </input-field>
  </div>
</body>
`;
